translator/trace/otlp/otlp_to_oc.go

```diff
@@ -0,0 +1,303 @@
+		}
+}
+
+func attributesToOC(attributes []*commonpb.AttributeKeyValue) map[string]string {
+	if len(attributes) == 0 {
+		return nil
+	}
+
+	ocAttributes := make(map[string]string, len(attributes))
+	for _, attr := range attributes {
+		ocAttributes[attr.Key] = attributeValueToString(attr)
+	}
+	return ocAttributes
+}
+
+func attributesToOCSpanAttributes(attributes []*commonpb.AttributeKeyValue, droppedCount uint32) *octrace.Span_Attributes {
+	if len(attributes) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_Attributes{
+		AttributeMap:           attributesToOCMap(attributes),
+		DroppedAttributesCount: int32(droppedCount),
+	}
+}
+
+func attributesToOCMap(attributes []*commonpb.AttributeKeyValue) map[string]*octrace.AttributeValue {
+	if len(attributes) == 0 {
+		return nil
+	}
+
+	ocAttributes := make(map[string]*octrace.AttributeValue, len(attributes))
+	for _, attr := range attributes {
+		ocAttributes[attr.Key] = attributeValueToOC(attr)
+	}
+	return ocAttributes
+}
+
+func attributeValueToString(attr *commonpb.AttributeKeyValue) string {
+	switch attr.Type {
+	case commonpb.AttributeKeyValue_STRING:
+		return attr.StringValue
+	case commonpb.AttributeKeyValue_BOOL:
+		return strconv.FormatBool(attr.BoolValue)
+	case commonpb.AttributeKeyValue_DOUBLE:
+		return strconv.FormatFloat(attr.DoubleValue, 'f', -1, 64)
+	case commonpb.AttributeKeyValue_INT:
+		return strconv.FormatInt(attr.IntValue, 10)
+	default:
+		return fmt.Sprintf("<Unknown OpenTelemetry attribute value type %q>", attr.Type)
+	}
+}
+
+func attributeValueToOC(attr *commonpb.AttributeKeyValue) *octrace.AttributeValue {
+	a := &octrace.AttributeValue{}
+
+	switch attr.Type {
+	case commonpb.AttributeKeyValue_STRING:
+		a.Value = &octrace.AttributeValue_StringValue{
+			StringValue: &octrace.TruncatableString{
+				Value: attr.StringValue,
+			},
+		}
+	case commonpb.AttributeKeyValue_BOOL:
+		a.Value = &octrace.AttributeValue_BoolValue{
+			BoolValue: attr.BoolValue,
+		}
+	case commonpb.AttributeKeyValue_DOUBLE:
+		a.Value = &octrace.AttributeValue_DoubleValue{
+			DoubleValue: attr.DoubleValue,
+		}
+	case commonpb.AttributeKeyValue_INT:
+		a.Value = &octrace.AttributeValue_IntValue{
+			IntValue: attr.IntValue,
+		}
+	default:
+		a.Value = &octrace.AttributeValue_StringValue{
+			StringValue: &octrace.TruncatableString{
+				Value: fmt.Sprintf("<Unknown OpenTelemetry attribute value type %q>", attr.Type),
+			},
+		}
+	}
+
+	return a
+}
+
+// OTLP follows the W3C format, e.g. "vendorname1=opaqueValue1,vendorname2=opaqueValue2"
+// See https://w3c.github.io/trace-context/#combined-header-value
+func tracestateToOC(tracestate string) *octrace.Span_Tracestate {
+	if tracestate == "" {
+		return nil
+	}
+
+	// key-value pairs in the "key1=value1" format
+	pairs := strings.Split(tracestate, ",")
+
+	entries := make([]*octrace.Span_Tracestate_Entry, 0, len(pairs))
+	for _, pair := range pairs {
+		kv := strings.Split(pair, "=")
```

Comment:
```suggestion
		kv := strings.SplitN(pair, "=", 2)
```
This is slightly better in case non-conformant sender uses the equal sign in the value.

Also do `if len(pair)>=2` to ensure we have 2 elements and handle the case when len==1 as well (val missing).